File: app/src/main/java/com/example/speedtest_rework/core/upload/Uploader.java
package com.example.speedtest_rework.core.upload;

import com.example.speedtest_rework.core.base.Connection;

import java.io.DataOutputStream;
import java.util.Random;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLSession;
import javax.net.ssl.SSLSocketFactory;


public abstract class Uploader extends Thread {
    private Connection c;
    private String path;
    private boolean stopASAP = false, resetASAP = false;
    private long totUploaded = 0;
    private byte[] garbage;

    public Uploader(Connection c, String path, int ckSize) {
        this.c = c;
        this.path = path;
        Random r = new Random(System.nanoTime());
        r.nextBytes(garbage);
        start();
    }

    byte[] buffer = new byte[150 * 1024];

    public void run() {
        long uploadedKByte = 0;

        while (true) {
            try {
                HttpsURLConnection conn = null;
                conn = (HttpsURLConnection) url.openConnection();
                conn.setDoOutput(true);
                conn.setRequestMethod("POST");
                conn.setRequestProperty("Connection", "Keep-Alive");
                conn.setSSLSocketFactory((SSLSocketFactory) SSLSocketFactory.getDefault());
                conn.setHostnameVerifier(new HostnameVerifier() {
                    @Override
                    public boolean verify(String hostname, SSLSession session) {
                        return true;
                    }
                });
                conn.connect();
                DataOutputStream dos = new DataOutputStream(conn.getOutputStream());


                dos.write(buffer, 0, buffer.length);
                dos.flush();

                conn.getResponseCode();

                uploadedKByte += buffer.length / 1024.0;
                long endTime = System.currentTimeMillis();
                double uploadElapsedTime = (endTime - startTime) / 1000.0;
                if (uploadElapsedTime >= timeout) {
                    break;
                }

                dos.close();
                conn.disconnect();
            } catch (Exception ex) {
                ex.printStackTrace();
                break;
            }
        }
    }


    public void stopASAP() {
        this.stopASAP = true;
    }

    public abstract void onProgress(long uploaded);

    public abstract void onError(String err);

    public void resetUploadCounter() {
        resetASAP = true;
    }

    public long getUploaded() {
        return resetASAP ? 0 : totUploaded;
    }
}

File: app/src/main/java/com/example/speedtest_rework/core/upload/Uploader.kt
package com.example.speedtest_rework.core.upload

import com.example.speedtest_rework.core.base.Connection
import kotlinx.coroutines.*
import java.util.*

abstract class Uploader(private val c: Connection, private val path: String, ckSize: Int) {
    private var stopASAP = false
    private var resetASAP = false
    private var totUploaded: Long = 0
    private val garbage: ByteArray
    private var job: Job? = null
    private var coroutineScope: CoroutineScope? = null

    init {
        garbage = ByteArray(ckSize * 1024)
        val r = Random(System.nanoTime())
        r.nextBytes(garbage)
    }

    fun startUpload() {
        coroutineScope = CoroutineScope(Dispatchers.IO)
        job = coroutineScope?.launch {
            try {
                val s = path
                var lastProgressEvent = System.currentTimeMillis()
                val out = c.outputStream
                while (true) {
                    if (stopASAP) break
                    c.POST(s, true, "application/octet-stream", garbage.size.toLong())
                    var offset = 0
                    while (offset < garbage.size) {
                        if (stopASAP) break
                        val l =
                            if (offset + BUFFER_SIZE >= garbage.size) garbage.size - offset else BUFFER_SIZE
                        out!!.write(garbage, offset, l)
                        if (stopASAP) break
                        if (resetASAP) {
                            totUploaded = 0
                            resetASAP = false
                        }
                        totUploaded += l.toLong()
                        if (System.currentTimeMillis() - lastProgressEvent > 100) {
                            lastProgressEvent = System.currentTimeMillis()
                            onProgress(totUploaded)
                        }
                        offset += BUFFER_SIZE
                    }
                    if (stopASAP) break
                    while (c.readLineUnbuffered()!!.trim { it <= ' ' }.isNotEmpty());
                }
                c.close()
            } catch (t: Throwable) {
                try {
                    c.close()
                } catch (t1: Throwable) {
                }
                onError(t.toString())
            }
        }

    }

    fun stopUpload() {
        coroutineScope?.cancel()
        job?.cancel()
        stopASAP = true
    }

    abstract fun onProgress(uploaded: Long)
    abstract fun onError(err: String?)
    fun resetUploadCounter() {
        resetASAP = true
    }

    val uploaded: Long
        get() = if (resetASAP) 0 else totUploaded

    companion object {
        private const val BUFFER_SIZE = 150 * 1024
    }
}